drivers/eeprom/eeprom_at2x.c

```diff
@@ -599,74 +597,172 @@ static const struct eeprom_driver_api eeprom_at2x_api = {
+	static const struct eeprom_at2x_config eeprom_##name##_config_##index = {                  \
+		.bus = bus_(index),                                                                \
+		EEPROM_AT2X_WP_GPIOS(index).size = DT_INST_PROP(index, size),                      \
+		.pagesize = DT_INST_PROP(index, pagesize),                                         \
+		.addr_width = DT_INST_PROP(index, address_width),                                  \
+		.readonly = DT_INST_PROP(index, read_only),                                        \
+		.timeout = DT_INST_PROP(index, timeout),                                           \
+		.bus_api = bus_api_,                                                               \
+	};                                                                                         \
+	static struct eeprom_at2x_data eeprom_##name##_data_##index;                               \
+	DEVICE_DT_INST_DEFINE(index, &eeprom_at2x_init, NULL, &eeprom_##name##_data_##index,       \
+			      &eeprom_##name##_config_##index, POST_KERNEL,                        \
+			      CONFIG_EEPROM_AT2X_INIT_PRIORITY, &eeprom_at2x_api)
+
+#define DT_DRV_COMPAT atmel_at24
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+static struct eeprom_at2x_bus_api atmel_at24_bus_api = {
+	.bus_is_ready = eeprom_at2x_i2c_bus_is_ready,
+	.read = eeprom_at2x_i2c_read,
+	.write = eeprom_at2x_i2c_write,
+};
+#define EEPROM_ATMEL_AT24_BUS_API &atmel_at24_bus_api
+DT_INST_FOREACH_STATUS_OKAY_VARGS(EEPROM_AT2X_INST_DEFINE, DT_DRV_COMPAT, EEPROM_AT2X_I2C_BUS,
+				  EEPROM_ATMEL_AT24_BUS_API, ASSERT_AT24_ADDR_W_VALID)
 #endif
-
-#ifdef CONFIG_EEPROM_AT25
-INST_DT_AT2X_FOREACH(25, EEPROM_AT25_DEVICE);
+#undef DT_DRV_COMPAT
+
+#define DT_DRV_COMPAT atmel_at25
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+static struct eeprom_at2x_bus_api atmel_at25_bus_api = {
+	.bus_is_ready = eeprom_at2x_spi_bus_is_ready,
+	.read = eeprom_at2x_spi_read,
+	.write = eeprom_at2x_spi_write,
+};
+#define EEPROM_ATMEL_AT25_BUS_API &atmel_at25_bus_api
+DT_INST_FOREACH_STATUS_OKAY_VARGS(EEPROM_AT2X_INST_DEFINE, DT_DRV_COMPAT, EEPROM_AT2X_SPI_BUS,
+				  EEPROM_ATMEL_AT25_BUS_API, ASSERT_AT25_ADDR_W_VALID)
+#endif
+#undef DT_DRV_COMPAT
+
+#define DT_DRV_COMPAT st_m24m02_a125
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+static struct eeprom_at2x_bus_api st_m24m02_a125_bus_api = {
+	.bus_is_ready = eeprom_at2x_i2c_bus_is_ready,
+	.read = eeprom_at2x_i2c_read,
+	.write = eeprom_at2x_i2c_write,
+};
+#define EEPROM_ST_M24M02_A125_BUS_API &st_m24m02_a125_bus_api
+DT_INST_FOREACH_STATUS_OKAY_VARGS(EEPROM_AT2X_INST_DEFINE, DT_DRV_COMPAT, EEPROM_AT2X_I2C_BUS,
+				  EEPROM_ST_M24M02_A125_BUS_API, ASSERT_AT24_ADDR_W_VALID)
+#endif
+#undef DT_DRV_COMPAT
+
+#define DT_DRV_COMPAT st_m24m01_a125
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+static struct eeprom_at2x_bus_api st_m24m01_a125_bus_api = {
+	.bus_is_ready = eeprom_at2x_i2c_bus_is_ready,
+	.read = eeprom_at2x_i2c_read,
+	.write = eeprom_at2x_i2c_write,
+};
```

Comment:
No need to create `api` instances for each compatible, just re-use a single `spi` or `i2c` instance.

Reply:
good point